File: matrix/incope2.hs
{-# OPTIONS -fglasgow-exts #-}

-- Interpreter, Compiler, Partial Evaluator
-- Inspired by the solution by Ken in incope

module Incope2 where

{-
  The language is simply-typed lambda-calculus with fixpoint,
  integers and comparison.

  Lam hoas_fn | App e e | Fix hoas_fn |
  I Int | Add ie1 ie2 |
  IFEQ ie1 ie2 e-then e-else
  
  The language is just expressive enough for the Gibonacci function.

  The compiler, the interpreter and the source and target languages
  are *all* typed. The interpreter and the compiler use no tags.
  There is no pattern-match failure possible: the evaluators never
  get stuck.

  The implementation below is *almost* possible in OCaml. Indeed,
  one can emulate Haskell typeclasses using explicit dictionary passing.
  The dictionary will by a polymorphic (rank-2) record -- but it is OK,
  OCaml permits it. Alas, The typeclass below uses higher-order type
  constructors: repr and res are of a kind * -> *. OCaml does not support
  higher-order polymorphism.

-}

-- This class defines syntax (and its instances, semantics) of our language

class Symantics rep where
    int :: Int -> rep Int Int               -- int literal
    add :: rep Int Int -> rep Int Int -> rep Int Int
    ifeq:: rep Int Int -> rep Int Int -> rep sa da -> rep sa da -> rep sa da
    fix :: (rep sa da -> rep sa da) -> rep sa da

    lam :: (rep sa da -> rep sb db) -> rep (rep sa da -> rep sb db) (da->db)
    app :: rep (rep sa da -> rep sb db) (da->db) -> rep sa da -> rep sb db


-- The following `projection' function is specific to repr.
-- It is like `run' of the monad
--    comp :: repr a -> something a

test1 () = add (int 1) (int 2)
test2 () = lam (\x -> add x x)
test3 () = lam (\x -> add (app x (int 1)) (int 2))

testgib () = lam (\x -> lam (\y ->
                  fix (\self -> lam (\n ->
                      ifeq n (int 0) x
                        (ifeq n (int 1) y
                         (add (app self (add n (int (-1))))
                              (app self (add n (int (-2))))))))))


testgib1 () = app (app (app (testgib ()) (int 1)) (int 1)) (int 5)

-- ------------------------------------------------------------------------
-- The interpreter
-- It is typed, tagless interpreter: R is not a tag. The interpreter
-- never gets stuck, because it evaluates typed terms only

-- Note that everything going on here is straight out of
-- "Boxes go Bananas" by Washburn and Weirich (intended or otherwise)
newtype R sa da = R da deriving Show
unR (R x) = x


instance Symantics R where
    int x = R x
    add e1 e2 = R( (unR e1) + (unR e2) )
    ifeq ie1 ie2 et ee = R( if (unR ie1) == (unR ie2) then unR et else unR ee )
    fix f = R( fx (unR . f . R)) where fx f = f (fx f)
    lam f = R (unR . f . R)
    app e1 e2 = R( (unR e1) (unR e2) )

mkitest f = unR (f $ ())

itest1 = mkitest test1
itest2 = mkitest test2
itest3 = mkitest test3
itestg = mkitest testgib1


-- ------------------------------------------------------------------------
-- The compiler
-- We compile to GADT, to be understood as a typed assembly language
-- (typed bytecode). The GADT does _not_ use the higher-order abstract
-- syntax. We could have used template Haskell. Alas, its expressions
-- are untyped.
-- Note how ByteCode represents MetaOCaml's `code'. Thus the compiler
-- below neatly maps to the MetaOCaml (with no GADTs).
-- Note how the compiler never raises any exception and matches no tags
-- (no generated code has any tags)

data ByteCode t where
    Var :: Int -> ByteCode t                -- variables identified by numbers
    Lam :: Int -> ByteCode t2 -> ByteCode (t1->t2)
    App :: ByteCode (t1->t2) -> ByteCode t1  -> ByteCode t2
    Fix :: Int -> ByteCode t -> ByteCode t
    INT :: Int -> ByteCode Int
    Add :: ByteCode Int -> ByteCode Int -> ByteCode Int
    IFEQ :: ByteCode Int -> ByteCode Int -> ByteCode t -> ByteCode t ->
            ByteCode t
instance Show (ByteCode t) where
    show (Var n) = "V" ++ show n
    show (Lam n b) = "(\\V" ++ show n ++ " -> " ++ show b ++ ")"
    show (App e1 e2) = "(" ++ show e1 ++ " " ++ show e2 ++ ")"
    show (Fix n b) = "(fix\\V" ++ show n ++ " " ++ show b ++ ")"
    show (INT n) = show n
    show (Add e1 e2) = "(" ++ show e1 ++ " + " ++ show e2 ++ ")"
    show (IFEQ ie1 ie2 et ee)
        = "(if " ++ show ie1 ++ " == " ++ show ie2  ++
          " then " ++ show et ++ " else " ++ show ee ++ ")"


-- Int is the variable counter
-- for allocation of fresh variables
newtype C st dt = C (Int -> (ByteCode dt, Int)) 
unC (C t) vc0 = t vc0

instance Symantics C where
    int x = C(\vc -> (INT x, vc))
    add e1 e2 = C(\vc -> let (e1b,vc1) = unC e1 vc
                             (e2b,vc2) = unC e2 vc1
                         in (Add e1b e2b,vc2))

    ifeq ie1 ie2 et ee = C(\vc -> let (ie1b,vc1) = unC ie1 vc
                                      (ie2b,vc2) = unC ie2 vc1
                                      (etb,vc3)  = unC et  vc2
                                      (eeb,vc4)  = unC ee  vc3
                         in (IFEQ ie1b ie2b etb eeb,vc4))
    fix f = C(\vc -> let v = vc
                         var = C(\vc -> (Var v, vc))
                         (body,vc') = unC (f var) (succ vc)
                     in (Fix v body, vc'))

    lam f = C(\vc -> let v = vc
                         var = C(\vc -> (Var v, vc))
                         (body,vc') = unC (f var) (succ vc)
                     in (Lam v body, vc'))
    app e1 e2 = C(\vc -> let (e1b,vc1) = unC e1 vc
                             (e2b,vc2) = unC e2 vc1
                         in (App e1b e2b,vc2))


compC repr = fst $ unC repr 0

ctest1 = compC . test1 $ ()
ctest2 = compC . test2 $ ()
ctest3 = compC . test3 $ ()
ctestg = compC . testgib1 $ ()


-- ------------------------------------------------------------------------
-- The partial evaluator

-- We need no Lift byte-code instruction: no parametric CSP. That is great!

-- The code below is NOT parametric. We could have used type-classes
-- instead of GADTs. Ken noted that the code below could be re-functionalized,
-- and so could in fact be translated into MetaOCaml.
-- Or, if we observe the correspondence between GADT and MetaOCaml <code>,
-- perhaps we can translate to MetaOCaml with an additional layer of code?

data P st dt = P (Maybe st) (C st dt)

cC :: C st1 dt -> C st2 dt
cC = C . unC

abstr :: P st dt -> C st dt
abstr (P _ e) = e

instance Symantics P where
    int x = P (Just x) (int x)
    add e1 e2 = case (e1,e2) of
                 (P (Just n1) _,P (Just n2) _) -> int (n1 + n2)
                 _ -> P Nothing $ add (abstr e1) (abstr e2)

    -- lam :: (rep sa da -> rep sb db) -> rep (rep sa da -> rep sb db) (da->db)
    lam f = P (Just f) (cC (lam (abstr . f . P Nothing)))
    -- app :: repr (a->b) -> repr a -> repr b
    app ef ea = case ef of
                        P (Just f) _ -> f ea
                        P _ f -> P Nothing (app (cC f) (abstr ea))

    ifeq ie1 ie2 et ee = case (ie1,ie2) of
                          (P (Just n1) _, P (Just n2) _) -> 
			      if n1==n2 then et else ee
                          _ -> P Nothing $ ifeq (abstr ie1) (abstr ie2)
                                           (abstr et) (abstr ee)
    -- fix :: (repr a -> repr a) -> repr a
    -- For now, to avoid divergence at the PE stage, we residualize
    -- actually, we unroll the fixpoint exactly once, and then
    -- residualize
    fix f = f (P Nothing (fix (abstr . f . P Nothing)))

compP x = compC . abstr $ x

ptest1 = compP . test1 $ ()
ptest2 = compP . test2 $ ()
ptest3 = compP . test3 $ ()

ptestg = compP . testgib1 $ ()

File: matrix/incope2.ml
(* Interpreter, Compiler, Partial Evaluator *)

(*
  The language is simply-typed lambda-calculus with fixpoint,
  integers [plus basic operations], booleans [ditto] and comparison.

  Lam hoas_fn | App e e | Fix hoas_fn |
  I Int | Add ie1 ie2 | Mul ie1 ie2 
  B Bool |
  IF b e-then e-else
  
  The language is just expressive enough for the Gibonacci function.

  The compiler, the interpreter and the source and target languages
  are *all* typed. The interpreter and the compiler use no tags.
  There is no pattern-match failure possible: the evaluators never
  get stuck.
*)


(* This is used only to extract the result so we can see it nicely.
   Alternatively, we can always extract the result as code,
   especially if we don't use the interpreter.
*)
type ('a,'v) result = RL of 'v | RC of ('a,'v) code;;

module type T = sig type ('a, 'b, 'c) repr end;;
module Q(T:T) = struct
  type ('c, 'sv, 'dv) pack = {
   int  : int  -> ('c,int,int) T.repr;
   bool : bool -> ('c,bool,bool) T.repr;
   add  : ('c,int,int) T.repr -> ('c,int,int) T.repr -> ('c,int,int) T.repr;
   mul  : ('c,int,int) T.repr -> ('c,int,int) T.repr -> ('c,int,int) T.repr;
   leq  : ('c,int,int) T.repr -> ('c,int,int) T.repr -> ('c,bool,bool) T.repr;
   eql  : ('c,'sv,'dv) T.repr -> ('c,'sv,'dv) T.repr -> ('c,bool,bool) T.repr;
   if_ : ('c,bool,bool) T.repr ->
             (unit -> ('c,'sv,'dv) T.repr) ->
             (unit -> ('c,'sv,'dv) T.repr) -> ('c,'sv,'dv) T.repr;
   lam : 'sa 'sb 'da 'db . (('c,'sa,'da) T.repr -> ('c,'sb,'db) T.repr)
    -> ('c,(('c,'sa,'da) T.repr -> ('c,'sb,'db) T.repr),'da->'db) T.repr;
   app : 'da 'db 'sa 'sb . 
    ('c,(('c,'sa,'da) T.repr -> ('c,'sb,'db) T.repr),'da->'db) T.repr
    -> ('c,'sa,'da) T.repr -> ('c,'sb,'db) T.repr;
   fix : 'sa 'sb 'da 'db . 
      (('c,(('c,'sa,'da) T.repr -> ('c,'sb,'db) T.repr) as 's,'da->'db) T.repr 
             -> ('c,'s,'da->'db) T.repr)  -> ('c,'s,'da->'db) T.repr
  }
end ;;

(* This class/type defines syntax (and its instances, semantics) 
   of our language
 *)

module type Symantics = sig
  type ('c,'sv,'dv) repr
  val int  : int  -> ('c,int,int) repr
  val bool : bool -> ('c,bool,bool) repr
  val add  : ('c,int,int) repr -> ('c,int,int) repr -> ('c,int,int) repr
  val mul  : ('c,int,int) repr -> ('c,int,int) repr -> ('c,int,int) repr
  val leq  : ('c,int,int) repr -> ('c,int,int) repr -> ('c,bool,bool) repr
  (* could be defined in terms of leq and if_ *)
  val eql  : ('c,'a,'a) repr -> ('c,'a,'a) repr -> ('c,bool,bool) repr
  (* The last two arguments to if are functional terms.
     One of them is applied to unit.
     The reason for this charade is to prevent evaluation
     of both arguments of if_ in a CBV language.
     if_ must be a syntactic form rather than just a function.
     Or, at least it shouldn't take just (undelayed) terms.
  *)
  val if_ : ('c,bool,bool) repr ->
             (unit -> ('c,'sa,'da) repr) ->
             (unit -> ('c,'sa,'da) repr) -> ('c,'sa,'da) repr 

  val lam : (('c,'sa,'da) repr -> ('c,'sb,'db) repr)
    -> ('c,(('c,'sa,'da) repr -> ('c,'sb,'db) repr),'da->'db) repr
  val app : ('c,(('c,'sa,'da) repr -> ('c,'sb,'db) repr),'da->'db) repr
    -> ('c,'sa,'da) repr -> ('c,'sb,'db) repr
  val fix : (('c,(('c,'sa,'da) repr -> ('c,'sb,'db) repr) as 's,'da->'db) repr 
             -> ('c,'s,'da->'db) repr)  -> ('c,'s,'da->'db) repr
(* Should we keep this in the comments? (JC: yes)
  val unfold : ('c,'s,'a) repr -> 
               ('c,('c,'s,'a) repr -> ('c,'s,'a) repr,'a->'a) repr -> 
               ('c, ('c,int,int) repr -> ('c,'s,'a) repr, int->'a) repr
*)
  val get_res : ('c,'sv,'dv) repr -> ('c,'dv) result

end
;;


(* Running example *)

module EX(S: Symantics) = struct
 module QQ = Q(S)
 open QQ
 let p = {
     int = S.int; bool = S.bool; add = S.add; mul = S.mul; leq = S.leq; 
     eql = S.eql; if_ = S.if_; lam = S.lam; app = S.app; fix = S.fix}
       
 (* Unit is to prevent monomorphising over 'c *)
 let test1 e = e.add (e.int 1) (e.int 2)
 let test2 e = e.lam (fun x -> e.add x x)
 let test3 e = e.lam (fun x -> e.add (e.app x (e.int 1)) (e.int 2))

 let testgib e = e.lam (fun x -> e.lam (fun y ->
                  e.fix (fun self -> e.lam (fun n ->
                      e.if_ (e.leq n (e.int 0)) (fun () -> x)
                        (fun () ->
                          (e.if_ (e.leq n (e.int 1)) (fun () -> y)
                           (fun () ->
                             (e.add (e.app self (e.add n (e.int (-1))))
                                    (e.app self (e.add n (e.int (-2))))))))))))

 let testgib1 e = e.app (e.app (e.app (testgib e) (e.int 1)) (e.int 1)) (e.int 5)
 let testgib2 e = e.lam (fun x -> (e.lam (fun y ->
   e.app (e.app (e.app (testgib e) x) y) (e.int 5))))

 let testpowfix e = e.lam (fun x ->
                      e.fix (fun self -> e.lam (fun n ->
                        e.if_ (e.leq n (e.int 0)) (fun () -> e.int 1)
                            (fun () -> e.mul x (e.app self (e.add n (e.int (-1))))))))

 let testpowfix7 e = e.lam (fun x -> e.app (e.app (testpowfix e) x) (e.int 7))

 let runit t = S.get_res (t p)

 let test1r = runit test1
 let test2r = runit test2
 let test3r = runit test3

 let testgibr = runit testgib
 let testgib1r = runit testgib1
 let testgib2r = runit testgib2

 let testpowfixr = runit testpowfix
 let testpowfix7r = runit testpowfix7

end;;


(* ------------------------------------------------------------------------ *)
(* The interpreter
   It is typed, tagless interpreter: R is not a tag. The interpreter
   never gets stuck, because it evaluates typed terms only
*)
(* Pure interpreter. It is essentially the identity transformer *)
module R = struct
  type ('c,'sv,'dv) repr = 'dv    (* absolutely no wrappers *)
  let int (x:int) = x
  let bool (b:bool) = b
  let add e1 e2 = e1 + e2
  let mul e1 e2 = e1 * e2
  let leq x y = x <= y
  let eql x y = x == y
  let if_ eb et ee = if eb then (et ()) else (ee ())

  let lam f = f
  let app e1 e2 = e1 e2
  let fix f n = let rec fx f n = f (fx f) n in fx f n
  let rec unfold z s = fun n -> if n<=0 then z else s ((unfold z s) (n-1))

  let get_res x = RL x
end;;

module EXR = EX(R);;
(* tests moved to the end, all together, to make comparisons easier *)

(* ------------------------------------------------------------------------ *)
(* Pure compiler *)

(* Note how the compiler never raises any exception and matches no tags
  (nor generated code has any tags)
*)

module C = struct
  type ('c,'sv,'dv) repr = ('c,'dv) code
  let int (x:int) = .<x>.
  let bool (b:bool) = .<b>.
  let add e1 e2 = .<.~e1 + .~e2>.
  let mul e1 e2 = .<.~e1 * .~e2>.
  let leq x y = .< .~x <= .~y >.
  let eql x y = .< .~x == .~y >.
  let if_ eb et ee = 
    .<if .~eb then .~(et () ) else .~(ee () )>.

  let lam f = .<fun x -> .~(f .<x>.)>.
  let app e1 e2 = .<.~e1 .~e2>.
  let fix f = .<fun n -> let rec self n = .~(f .<self>.) n in self n>.
  let unfold z s = .<let rec f n = if n <= 0 then .~z else .~s (f (n-1)) in f>.

  let get_res x = RC x
  let dyn_id (x : ('c,'sv,'dv) repr) : ('c,'sv1,'dv) repr = x
end;;

module EXC = EX(C);;

(* ------------------------------------------------------------------------ *)
(* Partial evaluator *)
(* Inspired by Ken's solution *)

module P =
struct
  type ('c,'sv,'dv) repr = {st: 'sv option; dy: ('c,'dv) code}
  let abstr {dy = x} = x
  let pdyn x = {st = None; dy = x}

  let int  (x:int)  = {st = Some (R.int x); dy = C.int x}
  let bool (x:bool) = {st = Some (R.bool x); dy = C.bool x}

  (* generic build - takes a repr constructor, an interpreter function
     and a compiler function (all binary) and builds a PE version *)
  let build cast f1 f2 = fun e1 -> fun e2 -> match (e1,e2) with
                   ({st = Some n1}, {st = Some n2}) -> cast (f1 n1 n2)
                 | _ -> pdyn (f2 (abstr e1) (abstr e2))
  (* same as 'build' but takes care of the neutral element (e) simplification
     allowed via a monoid structure which is implicitly present *)
  let buildsimp cast e f1 f2 = fun e1 e2 -> match (e1,e2) with
                 | ({st = Some e'}, _) when e = e' -> e2
                 | (_, {st = Some e'}) when e = e' -> e1
                 | ({st = Some n1}, {st = Some n2}) -> cast (f1 n1 n2)
                 | _ -> pdyn (f2 (abstr e1) (abstr e2))

  let add e1 e2 = buildsimp int 0 R.add C.add e1 e2
  let mul e1 e2 = buildsimp int 1 R.mul C.mul e1 e2
  let leq e1 e2 = build bool R.leq C.leq e1 e2
  let eql e1 e2 = build bool R.eql C.eql e1 e2
  let if_ eb et ee = match eb with
                       {st = Some b} -> if b then et () else ee ()
                     | _ -> pdyn (C.if_ (abstr eb) 
                                      (fun () -> abstr (et ()))
                                      (fun () -> abstr (ee ())))

  let lam f = {st = Some f; 
               dy = C.lam (fun x -> abstr (f (pdyn x)))}

  let app ef ea = match ef with
                    {st = Some f} -> f ea
                  | _ -> pdyn (C.app (C.dyn_id (abstr ef)) (abstr ea))

   (*
     For now, to avoid divergence at the PE stage, we residualize
    actually, we unroll the fixpoint exactly once, and then
    residualize
   *)
  let fix f = f (pdyn (C.fix (fun x -> abstr (f (pdyn x)))))
  (* this should allow us controlled unfolding *)
  let unfold z s = lam (function
      | {st = Some n} -> 
              let rec f k = if k<=0 then z else
                                match s with
                                | {st = Some m} -> m (f (k-1))
                                | {dy = y}      -> pdyn (C.app y (abstr (f (k-1))))
              in f n
      | {dy = y}      -> pdyn (C.app (C.unfold (abstr z) (abstr s)) y))

  let get_res x = C.get_res (abstr x)
end;;

module P1 =
struct
  include P
(*
  type ('c,'sv,'dv) repr = {st: 'sv option; dy: ('c,'dv) code}
  let abstr {dy = x} = x
  let pdyn x = {st = None; dy = x}
*)
  let fix f = 
    let fdyn = C.fix (fun x -> abstr (f (pdyn x))) in
    let fdynn e = pdyn (C.app fdyn e.dy) in
    {st = Some (function {st = Some _} as e -> 
                  let rec self n = 
                    (match n.st with Some _ -> app (f (lam self)) n 
                                     | _ ->  fdynn n) in 
                  app (f (lam self)) e
               | e  -> fdynn e);
               dy = fdyn }
end;;


module EXP = EX(P1);;

(* all the tests together *)
let itest1 = EXR.test1r;;
let ctest1 = EXC.test1r;;
let ptest1 = EXP.test1r;;

let itest2 = EXR.test2r;;
let ctest2 = EXC.test2r;;
let ptest2 = EXP.test2r;;

let itest3 = EXR.test3r;;
let ctest3 = EXC.test3r;;
let ptest3 = EXP.test3r;;

let itestg = EXR.testgibr;;
let ctestg = EXC.testgibr;;
let ptestg = EXP.testgibr;;

let itestg1 = EXR.testgib1r;;
let ctestg1 = EXC.testgib1r;;
let ptestg1 = EXP.testgib1r;;

let itestg2 = EXR.testgib2r;;
let ctestg2 = EXC.testgib2r;;
let ptestg2 = EXP.testgib2r;;

let itestp7 = EXR.testpowfix7r;;
let ctestp7 = EXC.testpowfix7r;;
let ptestp7 = EXP.testpowfix7r;;

(* actually run some of the above tests *)
let ctest2' = let res = match (ctest2) with
    | (RL x) -> failwith "not what we want"
    | (RC x) -> (.< .~x 5 >.) 
in .! res;;
let ctest3' = let f = (fun x -> x+17) in
    let res = match ctest3 with
    | (RL x) -> failwith "not what we want"
    | (RC x) -> (.< .~x f >.) 
in .! res;;
let ctestg1' = 
    let res = match ctestg1 with
    | (RL x) -> failwith "not what we want"
    | (RC x) -> x
in .! res;;

(* That's all folks. It seems to work... *)


(* Remnants of an earlier idea: compile the PE: make a code, which,
when run, will make a PE...
*)


(*
module P = struct
  let abstr = function (RL x) -> .<x>. | RC x -> x
  let int (x:int) = .<RL x>.
  let add e1 e2 = .<
    match (e1,e2) with
      (RL n1, RL n2) -> RL (n1+n2)
    | _              -> RC (C.add (abstr e1) (abstr e2))>.
(*
  let ifeq ie1 ie2 et ee = 
    .<let i1 = .~ie1 in if i1 = .~ie2 then .~et i1 else .~ee i1>.

  let lam f = .<fun x -> .~(f .<x>.)>.
  let app e1 e2 = .<.~e1 .~e2>.
  let fix f = .<fun n -> let rec self n = .~(f .<self>.) n in self n>.
  let get_res x = .! x
*)
end;;

*)


(* Extension of S for an imperative language with a single piece of
   state.

        let x = e1 in e2
        deref e
        set e1 e2 (returning the old value of e1)
        The optional 
                begin e1; e2 end
        is just 
                let dummy=e1 in e2

 Since we use higher-order abstract syntax, let x = e1 in e2
 is just lapp e1 (\x -> e2), which is an inverse application.
 Some may call it `bind'.
*)

(* Extension of S for an imperative language

        let x = e1 in e2
        newref e
        deref e
        set e1 e2 (returning the old value of e1)
        The optional 
                begin e1; e2 end
        is just 
                let dummy=e1 in e2

 Since we use higher-order abstract syntax, let x = e1 in e2
 is just lapp e1 (\x -> e2), which is an inverse application.

*)
(*
module type SI = sig
  include S
  val lapp : ('c,'sa,'da) repr ->
    ('c,(('c,'sa,'da) repr -> ('c,'sb,'db) repr),'da->'db) repr
    -> ('c,'sb,'db) repr
  val 
*)

(* The following shows that splitting a type variable doesn't
   help for module subtyping. A split (constrained) type variable
   is less polymorphic, so it won't pass where the signature demands
   unconstrained variable.

module type S1 = sig
  type ('a,'v) rep
  val vi : int -> ('a,int) rep
end;;

module R1 = struct
  type ('a,'v) rep = ('b, ('v->'w) -> 'w) code constraint 'a = 'b * 'w
  let vi (x:int) = .<fun k -> k x>.
end;;

module E1(S:S1) = struct
  let t () = S.vi 1
end;;

module E11 = E1(R1);;
*)
